# Chapter 1 Exercises
## A Brief Introduction to Information Theory

In [59]:
import numpy as np

### <font color='blue'>1.1</font> *How many bits are necessary to represent the alphabet using a binary code if we only allow uppercase characters? How about if we allow both uppercase and lowercase characters?*

In [65]:
bit_upp = int(round(np.log2(26)))
bit_upp_low = int(round(np.log2(2*26)))
print(bit_upp, bit_upp_low) 

5 6


### <font color='green'>Answer: 5 bits for uppercase, 6 bits for upppercase and lowercase.</font>

### <font color='blue'>1.2</font> *Describe how you can create an OR gate using NOT gates and AND gates.*

### <font color='green'>Answer: OR(a,b) = NOT( AND( NOT(a), NOT(b) ) ) </font>

### <font color='blue'>1.3</font> *A kilobyte is 1024 bytes. How many messages can it store?*

In [69]:
n = 8*1024 #8 kilobits = 1 kilobyte
print(n)

8192


### <font color='green'>Answer: A kilobyte stores $2^{8192}$ messages</font>

### <font color='blue'>1.4</font> *What is the entropy associated with the tossing of a fair coin?*

In [70]:
def H(P):
    return -P*np.log2(P) #Equ 1.4
heads = H(0.5)
tails = H(0.5)
print(heads+tails)

1.0


### <font color='green'>Answer: The Shannon entropy equals 1</font>

### <font color='blue'>1.5</font> *Suppose that X consists of the characters A, B, C, D that occur in a signal with respective probabilities 0.1, 0.4, 0.25, and 0.25. What is the Shannon entropy?*

In [74]:
A,B,C,D = H(0.1),H(0.4),H(0.25),H(0.25)
print(A+B+C+D)

1.860964047443681


### <font color='green'>Answer: The Shannon entropy equals 1.861</font>

### <font color='blue'>1.6</font> *A room full of people has incomes distributed in the following way:* 
### n(25.5) = 3,  n(30) = 5,  n(42) = 7,  n(50) = 3,  n(63) = 1,  n(75) = 2,  n(90) = 1 
### *What is the most probable income? What is the average income? What is the variance of this distribution?*

In [64]:
income = [25.5,30,42,50,63,75,90]
people = [3,5,7,3,1,2,1]
total_people = np.sum(people)
probs = people/total_people

def exp_val(j,p):
    for i in range (0,6):
        exp_val = np.sum(j[i]*p[i])
    return exp_val

def var(j,p):
    var = exp_val(np.square(j),np.square(p)) - (exp_val(j,p))**2
    return var

print('The most probable income is 30 with probability',max(people)/total_people)
print('The average income is',(25.5*3 + 30*5 + 42*7 + 50*3 + 63*1 + 75*2 + 90*1)/total_people)
print('The variance is',var(income, probs)) #Shouldn't be zero.

The most probable income is 30 with probability 0.3181818181818182
The average income is 44.25
The variance is 0.0
